README.md
# slimenodes
(minetest mod)\
Make bouncy solid slime and non-flowing liquid slime in minetest!

Changes in poikilos fork: see CHANGELOG.md

## Authors
- [original version](https://forum.minetest.org/viewtopic.php?t=10423)
  (`https://repo.or.cz/minetest_slimenodes.git`):
  Wuzzy
- sound:
  Wuzzy
- bucket graphics:
  poikilos (Jake Gustafson), ElementW (CC BY-SA 3.0)
- all other graphics:
  Wuzzy

## License
code: see included [license.txt](https://github.com/poikilos/slimenodes/blob/master/license.txt)
media: [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/) poikilos and ElementW


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/).

## [git] - 2019-07-02
(poikilos)
### Added
- Add craftitems (8 Slime + Empty Bucket = 1 Liquid Slime Bucket).
- Use high-res bucket image generation if available (from Bucket_Game).

### Changed
- Rename "slimenodes_slime_solid_*.png" to
  "slimenodes_block_solid_*.png".
- Put everything in loops to make code more maintainable (less
  redundant).
- Make slime bucket cost more (7) glue to craft.

## [git] - 2018-02-16
(poikilos)
### Added
- Add physics.
  - Solid is bouncy.
  - All bouncy slime nodes stop fall damage (-100%; not sure how with
    liquid slime--engine ignores damage reduction for liquid at high
    speeds).
  - Liquid is liquid-like.
- Make buckets for all types so collectors are not needed (and so
  storage is nerfed to one per inventory slot).
- Make recipes for all types (only if you have `mesecons_materials:glue`
  or `technic:glue` since the recipe requires glue).

### Changed
- Make all liquid blocks behave like each other and all solid blocks
  behave like each other (though levels differ).
- Change sounds so liquid and solid each use a different one of the dug
  sounds. This doesn't seem to work unless using slime collectors.
- Slime collectors cannot be crafted since holding stacks of slime would
  allow overpowered griefing. They can still be given administratively,
  but in multiplayer they should only be used for cleanup so they aren't
  traded or otherwise abused later.


buckets.lua
-- Minetest 0.4 mod: slime bucket
-- See README.txt for licensing and other information.

-- minetest.register_alias("bucket_slime_solid_cyan", "slimenodes:bucket_slime_solid_cyan")
-- minetest.register_alias("bucket_slime_solid_green", "slimenodes:bucket_slime_solid_green")
-- minetest.register_alias("bucket_slime_solid_orange", "slimenodes:bucket_slime_solid_orange")
-- minetest.register_alias("bucket_slime_solid_violet", "slimenodes:bucket_slime_solid_violet")

-- Register a new liquid
--    source = name of the source node
--    flowing = name of the flowing node
--    itemname = name of the new bucket item (or nil if liquid is not takeable)
--    inventory_image = texture of the new bucket item (ignored if itemname == nil)
--    name = text description of the bucket item
--    groups = (optional) groups of the bucket item, for example {water_bucket = 1}
--    force_renew = (optional) bool. Force the liquid source to renew if it has a
--                  source neighbour, even if defined as 'liquid_renewable = false'.
--                  Needed to avoid creating holes in sloping rivers.
-- This function can be called from any mod (that depends on bucket).
-- - OldCoder (Bucket_Game)


-- craftitems.lua defines slimenodesColorNames
if rawget(_G, "bucket") and bucket.register_bucket_hash then
	for index = 1, #slimenodesColorNames do
		local descPart = slimenodesColorNames[index]
		local namePart = descPart:lower()
		local liquid_bh = {}
		liquid_bh.bucket_node = "slimenodes:bucket_slime_liquid_" .. namePart
		liquid_bh.desc = "Liquid " .. descPart .. " Slime Bucket"
		liquid_bh.source_node = "slimenodes:slime_liquid_" .. namePart
		liquid_bh.flowing_node = "slimenodes:slime_liquid_" .. namePart -- .. "_slope"
		liquid_bh.force_renew = false
		-- groups =  -- TODO?
		-- liquid_bh.inventory_image =  "slimenodes_block_liquid_" .. namePart .. ".png"
		liquid_bh.invsource_image =  "slimenodes_block_liquid_" .. namePart .. ".png"
		bucket.register_bucket_hash(liquid_bh)
		-- rawget(_G, "ocutil") and bucket.register_node_bucketable then
		--ocutil.register_node_bucketable ("slimenodes:bucket_slime_solid_" .. namePart, {
			--description = "Solid " .. descPart .. " Slime Bucket" ,
			--bucket_base = "slimenodes:slime_solid_" .. namePart ,
			--tiles = {"slimenodes_block_solid_" .. namePart .. ".png"} ,
			--groups = {fall_damage_add_percent=-100, disable_jump=1, puts_out_fire=1, cools_lava=1, liquid=3, slimey=3} , -- and levels? see init.lua
			--sounds = {
				--dug = {name="slimenodes_dig", gain=0.6},
				--place = {name="slimenodes_place", gain=0.6},
			--},
		--})
		local solid_bh = {}
		solid_bh.bucket_node = "slimenodes:bucket_slime_solid_" .. namePart
		solid_bh.desc = "Solid " .. descPart .. " Slime Bucket"
		solid_bh.source_node = "slimenodes:slime_solid_" .. namePart
		solid_bh.flowing_node = "slimenodes:slime_solid_" .. namePart -- .. "_slope"
		solid_bh.force_renew = false
		-- groups =  -- TODO?
		-- solid_bh.inventory_image = "slimenodes_block_solid_" .. namePart .. ".png"
		solid_bh.invsource_image = "slimenodes_block_solid_" .. namePart .. ".png"
		bucket.register_bucket_hash(solid_bh)
	end
elseif rawget(_G, "bucket") and bucket.register_liquid then
	-- source, flowing, itemname, inventory image, name, groups, force_renew
	for index = 1, #slimenodesColorNames do
		local descPart = slimenodesColorNames[index]
		local namePart = descPart:lower()
		bucket.register_liquid(
			"slimenodes:slime_liquid_" .. namePart,
			"slimenodes:slime_liquid_" .. namePart,
			"slimenodes:bucket_slime_liquid_" .. namePart,
			"slimenodes_bucket_slime_liquid_" .. namePart .. ".png",
			"Liquid "..descPart.." Slime Bucket"
		)
		bucket.register_liquid(
			"slimenodes:slime_solid_" .. namePart,
			"slimenodes:slime_solid_" .. namePart,
			"slimenodes:bucket_slime_solid_" .. namePart,
			"slimenodes_bucket_slime_solid_" .. namePart .. ".png",
			"Solid "..descPart.." Slime Bucket"
		)
	end
end


recipes.lua

-- craftitems.lua defines slimenodesColorNames
for index = 1, #slimenodesColorNames do
	local descPart = slimenodesColorNames[index]
	local namePart = descPart:lower()
	minetest.register_craft({
		type = "cooking",
		cooktime = 10,
		output = "slimenodes:bucket_slime_solid_" .. namePart,
		recipe = "slimenodes:bucket_slime_liquid_" .. namePart
	})
end

local mesecons_materials = minetest.get_modpath("mesecons_materials")
local technic = minetest.get_modpath("technic")

local glue_s = nil

if mesecons_materials then
	glue_s = "mesecons_materials:glue"
elseif technic then
	glue_s = "technic:glue"
end
if glue_s then
	-- craftitems.lua defines slimenodesColorNames
	for index = 1, #slimenodesColorNames do
		local descPart = slimenodesColorNames[index]
		local namePart = descPart:lower()
		local thisDye = "dye:" .. namePart
		if minetest.registered_items[thisDye] then
			minetest.register_craft({
				type = "shapeless",
				output = "slimenodes:bucket_slime_liquid_" .. namePart,
				recipe = {
					glue_s, glue_s, glue_s,
					glue_s, thisDye, glue_s,
					glue_s, "bucket:bucket_water", glue_s,
				},
			})
		end
	end
end

for index = 1, #slimenodesColorNames do
	local descPart = slimenodesColorNames[index]
	local namePart = descPart:lower()
	local frag = "slimenodes:slime_" .. namePart
	local bucketName = "slimenodes:bucket_slime_liquid_" .. namePart
	minetest.register_craft({
		type = "shapeless",
		output = bucketName,
		recipe = {
			frag, frag, frag,
			frag, frag, frag,
			frag, "bucket:bucket_empty", frag,
		},
	})
	if technic and technic.register_separating_recipe then
		technic.register_separating_recipe({
			input = {bucketName.." 1"},
			output = {frag..' 8', "bucket:bucket_empty 1"},
		})
	else
		minetest.register_craft({
			recipe = {
				{bucketName},
			},
			output = frag..' 8',
		})
	end
end


init.lua
-- Boilerplate to support localized strings if intllib mod is installed.
local S
if minetest.get_modpath("intllib") then
	S = intllib.Getter()
else
	S = function(s) return s end
end

slimenodesColorNames = { "Cyan", "Green", "Orange", "Violet" }
slimenodesParams = {}
slimenodesParams["cyan"] = {}
slimenodesParams["cyan"].alpha = 120
slimenodesParams["cyan"].liquid_viscosity = 1
slimenodesParams["cyan"].post_effect_color = {a=16, r=0, g=255, b=255}
slimenodesParams["green"] = {}
slimenodesParams["green"].alpha = 170
slimenodesParams["green"].liquid_viscosity = 2
slimenodesParams["green"].post_effect_color = {a=64, r=100, g=200, b=100}
slimenodesParams["orange"] = {}
slimenodesParams["orange"].alpha = 200
slimenodesParams["orange"].liquid_viscosity = 4
slimenodesParams["orange"].post_effect_color = {a=96, r=200, g=100, b=0}
slimenodesParams["violet"] = {}
slimenodesParams["violet"].alpha = 220
slimenodesParams["violet"].liquid_viscosity = 6
slimenodesParams["violet"].post_effect_color = {a=128, r=50, g=0, b=200}

if not deepcopy then
	-- See http://lua-users.org/wiki/CopyTable
	function deepcopy(orig)
		local orig_type = type(orig)
		local copy
		if orig_type == 'table' then
			copy = {}
			for orig_key, orig_value in next, orig, nil do
				copy[deepcopy(orig_key)] = deepcopy(orig_value)
			end
			setmetatable(copy, deepcopy(getmetatable(orig)))
		else -- number, string, boolean, etc
			copy = orig
		end
		return copy
	end
end

if not level1copy then
	-- See http://lua-users.org/wiki/CopyTable
	function level1copy(orig)
		local orig_type = type(orig)
		local copy
		if orig_type == 'table' then
			copy = {}
			for orig_key, orig_value in next, orig, nil do
				copy[orig_key] = orig_value
			end
		else -- number, string, boolean, etc
			copy = orig
		end
		return copy
	end
end

dofile(minetest.get_modpath("slimenodes").."/craftitems.lua")


-- craftitems.lua defines slimenodesColorNames
local slimeNodeLevel = 0
local slimenodesLiqDef = nil
for index = 1, #slimenodesColorNames do
	descPart = slimenodesColorNames[index]
	namePart = descPart:lower()
	local snLName = "slimenodes:slime_liquid_" .. namePart
	local snSName = "slimenodes:slime_solid_" .. namePart
	slimenodesLiqDef = {
		drawtype = "liquid",
		alpha = 120,
		sunlight_propagates = true,
		paramtype = "light",
		walkable = false,
		pointable = false,
		diggable = true,
		buildable_to = false,
		liquidtype = "source",
		liquid_viscosity = 1,
		liquid_range = 0,
		liquid_renewable = false,
		liquids_pointable = true,
		drowning = 0,
		sounds = {
			dug = {name="slimenodes_dig", gain=0.6},
			place = {name="slimenodes_place", gain=0.6},
		}
	}
	if nil then   -- rawget(_G, "default") and default.register_water then
		print("slimenodes is using default.register_water...")
		-- Liquid:
		local def = {
			alpha             = slimenodesLiqDef.alpha,
			basename          = "slimenodes:slime_liquid_" .. namePart,
			light             = -1             ,
			liquid_viscosity  = slimenodesLiqDef.liquid_viscosity,
			liquid_range      = slimenodesLiqDef.liquid_range,
			sounds            = slimenodesLiqDef.sounds,
			drowning          = slimenodesLiqDef.drowning,
		}
		if slimenodesParams[namePart]
				and slimenodesParams[namePart].post_effect_color then
			def.post_effect_color = slimenodesParams[namePart].post_effect_color
		else
			def.post_effect_color = nil
		end
		default.register_water(def)

		-- Solid:
		local def = {
			alpha             = slimenodesLiqDef.alpha,
			basename          = snSName,
			light             = -1             ,
			liquid_range      = 0,
			sounds            = slimenodesLiqDef.sounds,
		}
		if slimenodesParams[namePart].post_effect_color then
			def.post_effect_color = slimenodesParams[namePart].post_effect_color
		end
		default.register_water(def)
	else
		print("slimenodes is using register_node...")
		-- Liquid:
		-- -- why does this commented version of the code cause a block that
		-- -- tries to flow but has CONTENT_IGNORE error?
		--local def = slimenodesLiqDef
		--def.description = S("Liquid " .. descPart .. " Slime Block")
		--def.inventory_image = minetest.inventorycube(
			--"slimenodes_block_liquid_" .. namePart .. ".png"
		--)
		--def.tiles = { "slimenodes_block_liquid_" .. namePart .. ".png" }
		--def.special_tiles = {
			--{
				--name = "slimenodes_block_liquid_" .. namePart .. ".png",
				--backface_culling = false
			--},
		--}
		--def.drop = snLName
		---- liquid_alternative_flowing = "slimenodes:slime_liquid_" .. namePart .. "_slope"
		--liquid_alternative_flowing = snLName
		--liquid_alternative_source = snLName
		--if slimenodesParams[namePart].post_effect_color then
			--def.post_effect_color = slimenodesParams[namePart].post_effect_color
		--end
		--if slimenodesParams[namePart].alpha then
			--def.alpha = slimenodesParams[namePart].alpha
		--end
		--if slimenodesParams[namePart].liquid_viscosity then
			--def.liquid_viscosity = slimenodesParams[namePart].liquid_viscosity
		--end
		--def.groups = {fall_damage_add_percent=-100, disable_jump=1, puts_out_fire=1, cools_lava=1, liquid=3, slimey=3}
		--if slimeNodeLevel > 0 then
			--def.groups.level = slimeNodeLevel
		--end
		--minetest.register_node(snLName, def)
		minetest.register_node(snLName, {
			description = S("Liquid " .. descPart .. " Slime Block"),
			inventory_image = minetest.inventorycube("slimenodes_block_liquid_" .. namePart .. ".png"),
			drawtype = "liquid",
			tiles = { "slimenodes_block_liquid_" .. namePart .. ".png" },
			special_tiles = {
				{
					name = "slimenodes_block_liquid_" .. namePart .. ".png",
					backface_culling = false
				},
			},
			alpha = slimenodesParams[namePart].alpha or 120,
			sunlight_propagates = true,
			paramtype = "light",
			walkable = false,
			pointable = false,
			diggable = true,
			buildable_to = false,
			drop = snLName,
			liquidtype = "source",
			liquid_alternative_flowing = snLName,
			liquid_alternative_source = snLName,
			liquid_viscosity = slimenodesParams[namePart].liquid_viscosity or 1,
			liquid_range = 0,
			liquid_renewable = false,
			liquids_pointable = true,
			drowning = 0,
			post_effect_color = deepcopy(slimenodesParams[namePart].post_effect_color) or {a=16, r=128, g=128, b=128},
			groups = {fall_damage_add_percent=-100, disable_jump=1, puts_out_fire=1, cools_lava=1, liquid=3, slimey=3},
			sounds = {
				dug = {name="slimenodes_dig", gain=0.6},
				place = {name="slimenodes_place", gain=0.6},
			}
		})


		-- Solid:
		local solidGroups = {bouncy=90, fall_damage_add_percent=-100, slimey=3,}
		if slimeNodeLevel > 0 then
			solidGroups.level = slimeNodeLevel
		end

		minetest.register_node(snSName, {
			description = S("Solid " .. descPart .. " Slime Block"),
			tiles = { "slimenodes_block_solid_" .. namePart .. ".png" },
			walkable = true,
			groups = solidGroups,
			sounds = {
				dug = {name="slimenodes_dug", gain=0.6},
				place = {name="slimenodes_place", gain=0.6},
				footstep = {name="slimenodes_step", gain=0.2},
			}
		})

	end
	if slimeNodeLevel <= 3 then
		slimeNodeLevel = slimeNodeLevel + 1
	end
end

minetest.register_tool("slimenodes:collector_lvl1", {
	description = S("Slime collector level 1"),
	inventory_image = "slimenodes_collector_lvl1.png",
	liquids_pointable = true,
	tool_capabilities = {
		full_punch_interval = 1,
		groupcaps = {
			slimey = { times = { [3] = 2 }, uses = 0, maxlevel = 1 }
		}
	}
})

minetest.register_tool("slimenodes:collector_lvl2", {
	description = S("Slime collector level 2"),
	inventory_image = "slimenodes_collector_lvl2.png",
	liquids_pointable = true,
	tool_capabilities = {
		full_punch_interval = 1,
		groupcaps = {
			slimey = { times = { [2] = 3, [3] = 1.5 }, uses = 0, maxlevel = 2 }
		}
	}
})

minetest.register_tool("slimenodes:collector_lvl3", {
	description = S("Slime collector level 3"),
	inventory_image = "slimenodes_collector_lvl3.png",
	liquids_pointable = true,
	tool_capabilities = {
		full_punch_interval = 1,
		groupcaps = {
			slimey = { times = { [1] = 4, [2] = 1.5, [3] = 0.75 }, uses = 0, maxlevel = 3 }
		}
	}
})

minetest.register_tool("slimenodes:collector_lvl4", {
	description = S("Slime collector level 4"),
	inventory_image = "slimenodes_collector_lvl4.png",
	liquids_pointable = true,
	tool_capabilities = {
		full_punch_interval = 0.9,
		groupcaps = {
			slimey = { times = { [1] = 1, [2] = 0.5, [3] = 0.1 }, uses = 0, maxlevel = 4 }
		}
	}
})

dofile(minetest.get_modpath("slimenodes").."/buckets.lua")
dofile(minetest.get_modpath("slimenodes").."/recipes.lua")

